<!DOCTYPE html>
<html lang="en-us">

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.7.0">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Sabrina Ward">

  
  
  
    
  
  <meta name="description" content="Find this notebook at https://github.com/jakobnissen/automa_tutorial
In bioinformatics, we have a saying:
 The first step of any bioinformatics project is to define a new file format, incompatible with all previous ones.
 The situation might not be quite as bad as the saying implies, but it is true that we have a lot of different file formats, representing the various kinds of data we work with.
For that reason, creating file parsers is a central task in bioinformatics, and has almost become a craft in itself.">

  
  <link rel="alternate" hreflang="en-us" href="https://BioJulia.github.io/post/automa1/">

  


  
  
  
  <meta name="theme-color" content="#2962ff">
  

  
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha256-+N4/V/SbAFiW1MPBCXnfnP9QSN3+Keu+NlB+0ev/YKQ=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.5.1/leaflet.css" integrity="sha256-SHMGCYmST46SoyGgo4YR/9AlK1vf3ff84Aq9yK4hdqM=" crossorigin="anonymous">
    

    

    
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/lazysizes/5.1.2/lazysizes.min.js" integrity="sha256-Md1qLToewPeKjfAHU1zyPwOutccPAm5tahnaw7Osw0A=" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    
      

      
      

      
    
      

      
      
        
      

      
    
      

      
      

      
    

  

  
  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat:400,700%7CRoboto:400,400italic,700%7CRoboto+Mono&display=swap">
  

  
  
  
  
  <link rel="stylesheet" href="/css/academic.css">

  




  


  

  <link rel="manifest" href="/index.webmanifest">
  <link rel="icon" type="image/png" href="/images/icon_huf2e28fc1e802707079b8e0ffee62f4dc_19447_32x32_fill_lanczos_center_2.png">
  <link rel="apple-touch-icon" type="image/png" href="/images/icon_huf2e28fc1e802707079b8e0ffee62f4dc_19447_192x192_fill_lanczos_center_2.png">

  <link rel="canonical" href="https://BioJulia.github.io/post/automa1/">

  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="BioJulia">
  <meta property="og:url" content="https://BioJulia.github.io/post/automa1/">
  <meta property="og:title" content="Tutorial to Automa: Part 1 | BioJulia">
  <meta property="og:description" content="Find this notebook at https://github.com/jakobnissen/automa_tutorial
In bioinformatics, we have a saying:
 The first step of any bioinformatics project is to define a new file format, incompatible with all previous ones.
 The situation might not be quite as bad as the saying implies, but it is true that we have a lot of different file formats, representing the various kinds of data we work with.
For that reason, creating file parsers is a central task in bioinformatics, and has almost become a craft in itself."><meta property="og:image" content="https://BioJulia.github.io/images/logo_huf2e28fc1e802707079b8e0ffee62f4dc_19447_300x300_fit_lanczos_2.png">
  <meta property="twitter:image" content="https://BioJulia.github.io/images/logo_huf2e28fc1e802707079b8e0ffee62f4dc_19447_300x300_fit_lanczos_2.png"><meta property="og:locale" content="en-us">
  
    
      <meta property="article:published_time" content="2020-08-30T16:28:09&#43;02:00">
    
    <meta property="article:modified_time" content="2020-08-30T16:28:09&#43;02:00">
  

  


    






  






<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://BioJulia.github.io/post/automa1/"
  },
  "headline": "Tutorial to Automa: Part 1",
  
  "datePublished": "2020-08-30T16:28:09+02:00",
  "dateModified": "2020-08-30T16:28:09+02:00",
  
  "author": {
    "@type": "Person",
    "name": "Jakob Nybo Nissen"
  },
  
  "publisher": {
    "@type": "Organization",
    "name": "BioJulia",
    "logo": {
      "@type": "ImageObject",
      "url": "img/https://BioJulia.github.io/"
    }
  },
  "description": "Find this notebook at https://github.com/jakobnissen/automa_tutorial\nIn bioinformatics, we have a saying:\n The first step of any bioinformatics project is to define a new file format, incompatible with all previous ones.\n The situation might not be quite as bad as the saying implies, but it is true that we have a lot of different file formats, representing the various kinds of data we work with.\nFor that reason, creating file parsers is a central task in bioinformatics, and has almost become a craft in itself."
}
</script>

  

  


  
  
  
  
  
    <script src="https://cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.1/cookieconsent.min.js" integrity="sha256-5VhCqFam2Cn+yjw61zbBNrbHVJ6SRydPeKopYlngbiQ=" crossorigin="anonymous"></script>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.1/cookieconsent.min.css" integrity="sha256-zQ0LblD/Af8vOppw18+2anxsuaz3pWYyVWi+bTvTH8Q=" crossorigin="anonymous">
  
  <script>
  window.addEventListener("load", function(){
    window.cookieconsent.initialise({
      "palette": {
        "popup": {
          "background": "#2962ff",
          "text": "#fff"
        },
        "button": {
          "background": "#fff",
          "text": "#2962ff"
        }
      },
      "theme": "classic",
      "content": {
        "message": "This website uses cookies to ensure you get the best experience on our website.",
        "dismiss": "Got it!",
        "link": "Learn more",
        "href": "https://www.cookiesandyou.com"
      }
    })});
  </script>



  





  <title>Tutorial to Automa: Part 1 | BioJulia</title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" >

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  






  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/images/logo_huf2e28fc1e802707079b8e0ffee62f4dc_19447_0x70_resize_lanczos_2.png" alt="BioJulia"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/images/logo_huf2e28fc1e802707079b8e0ffee62f4dc_19447_0x70_resize_lanczos_2.png" alt="BioJulia"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#hero"><span>Home</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        
        
        
        
        
        

        <li class="nav-item">
          <a class="nav-link " href="https://juliahub.com/ui/Packages?q=BioJulia" target="_blank" rel="noopener"><span>Software Packages</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#posts"><span>Posts</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#people"><span>People</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#community"><span>Community</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#support"><span>Support BioJulia</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#related-organizations"><span>Related Organizations</span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      <li class="nav-item">
        <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
      </li>
      

      

    </ul>

  </div>
</nav>


  <article class="article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1>Tutorial to Automa: Part 1</h1>

  

  
    


<div class="article-metadata">

  
  
  
  
  <div>
    



  
  <span><a href="/authors/jakobnissen/">Jakob Nybo Nissen</a></span>

  </div>
  
  

  
  <span class="article-date">
    
    
      
    
    Aug 30, 2020
  </span>
  

  

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    24 min read
  </span>
  

  
  
  
  <span class="middot-divider"></span>
  <a href="/post/automa1/#disqus_thread"></a>
  

  
  

</div>

    














  
</div>



  <div class="article-container">

    <div class="article-style">
      <p><strong>Find this notebook at <a href="https://github.com/jakobnissen/automa_tutorial">https://github.com/jakobnissen/automa_tutorial</a></strong></p>
<p>In bioinformatics, we have a saying:</p>
<blockquote>
<p>The first step of any bioinformatics project is to define a new file format, incompatible with all previous ones.</p>
</blockquote>
<p>The situation might not be <em>quite</em> as bad as the saying implies, but it <em>is</em> true that we have a <em>lot</em> of different file formats, representing the various kinds of data we work with.</p>
<p>For that reason, creating file parsers is a central task in bioinformatics, and has almost become <a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec14">a craft in itself.</a> An awful, awful craft. For anything but the simplest formats, designing robust parsers is hard, and it is difficult to know if a parser you've built is solid. Insignificant details like trailing whitespace can break a seemingly well-built parser.</p>
<p>At its core, a parser is a program that checks whether some input data comforms to some format. For bioinformatics, we typically also want to load the data in the file into some data structure to work on it. The format itself is typically specified in some kind of higher-level, but unambiguous language, like <a href="https://en.wikipedia.org/wiki/Newick_format#Grammar">this description of the Newick format.</a> A parser then, can be viewed as a program that:</p>
<ul>
<li>
<p>Is given a format specified in a high-level &ldquo;language&rdquo;, e.g. &ldquo;a leaf in the Newick format consist of only a <em>name</em>, which is formed from the alphabet &ldquo;a-z, A-Z and _&quot;&quot;, and</p>
</li>
<li>
<p>Checks whether some input follows that format using relatively low-level instructions, e.g. checking whether the following bytes until the next <code>(</code>, <code>)</code>, <code>,</code> or <code>;</code> match the pattern <code>[A-Za-z_]</code></p>
</li>
</ul>
<p>The central job of a parser is then to act as a <em>translator</em> between different abstraction levels: The format is specified in a high-level language, but the actual data processing must happen at a lower level language.</p>
<p>When described like that, parser generation seems quite analogous to code compilation, and begins too look like a <em>machine's</em> job, not a human's. Machines are fast and accurate, and do not make oversights or mistakes - exactly the characteristics you want when making a parser. Let's leave the machine tasks to the machines; we humans ought to put our effort into what we do best: Specifying the formats themselves.</p>
<p>In this tutorial, I will introduce Automa, a Julia package for creating parsers. Actually, when reading this you might be able to tell that Automa solves a problem slightly more general than just creating parsers, but creating parsers is what we will use it for in this tutorial. Automa is not an easy package to use; it is complex and a little opaque, but it's worth all the effort and more. Parsers generated by Automa have several advantages over human-made parsers:</p>
<ul>
<li>They are failsafe: Any deviation of the input from the specified format, no matter how trivial, will cause it to fail. Therefore, they are much more reliant than human-written parsers.</li>
<li>They are easy to change. You can often easily change details in the input format, such as allowing whitespace, with minimal changes to your code,</li>
<li>They are quite fast, with Automa-code often being faster than even optimized human-crafted parsers.</li>
</ul>
<p>In this first part of the tutorial I will show how to create parsers from data loaded into memory in order to highlight Automa itself. In the second part, which I will release at a later date, I will show how to create a proper BioJulia reader and writer for a file format.</p>
<h2 id="our-format">Our format</h2>
<p>For our example, let us begin with a simlified version of the FASTA format - we can call it &ldquo;SimpleFasta&rdquo;. Our format will be defined as something that looks like this:</p>
<pre><code>&gt;first
TGATAGTAGTCGTGTGATA
&gt;second
AGGACCCAATTATCGGGGTAA
</code></pre>
<p>I.e.</p>
<ul>
<li>A SimpleFasta file is a sequence of zero or more <em>records</em></li>
<li>A <em>record</em> consists of <em>header</em> * newline * <em>sequence</em> * newline, where &ldquo;*&rdquo; signifies concatenation</li>
<li>A <em>header</em> consists of <code>&gt;</code> followed by one or more characters from the alphabet a-z</li>
<li>A <em>sequence</em> consists of one or more chatacters from the alphabet A, C, G or T.</li>
</ul>
<p>We can visualize a hypothetical SimpleFasta parser using a simple flowchart - note: You need to have <code>graphviz</code> installed to be able to run the code below.</p>
<pre><code class="language-julia">s = &quot;&quot;&quot;digraph {
graph [ rankdir = LR ];
begin [ shape = circle ];
begin -&gt; header;
header [ shape = circle ];
header -&gt; sequence;
sequence [ shape = circle ];
sequence -&gt; end;
end [ shape = circle ];
sequence -&gt; header;
begin -&gt; end;
}
&quot;&quot;&quot;
function display_machine()
    write(&quot;/tmp/fasta_machine.dot&quot;, s)
    run(`dot -Tsvg -o /tmp/fasta_machine.svg /tmp/fasta_machine.dot`)
    open(&quot;/tmp/fasta_machine.svg&quot;) do file
        display(&quot;image/svg+xml&quot;, String(read(file)))
    end
end;
</code></pre>
<pre><code class="language-julia">display_machine()
</code></pre>
<p><img src="automa_tutorial_2_0.svg" alt="svg"></p>
<p>After the beginning of the file, we expect to see a header. Next, we move to the sequence. From there, we can either reach the end of the file, or loop back to another header. Alternatively, in the case of an empty file, we may move from the beginning of the file to the end immediately without any headers or sequences. We consider an empty file to also be specification-compliant.</p>
<p>Note that this simple flowchart describes any <em>valid</em> SimpleFasta format. But if, for example, a file had two consecutive header lines, there is no path we can take through the flowchart from &ldquo;begin&rdquo; to &ldquo;end&rdquo; which describes the file, and we can tell that the input did not conform to the format.</p>
<p>The circles marked &ldquo;begin&rdquo;, &ldquo;header&rdquo;, &ldquo;sequence&rdquo; or &ldquo;end&rdquo; are what we call the four <em>states</em>. Because the parser shown in the diagram has a fixed list of possible states that it changes between, we call the parser a <em>finite state machine</em> (FSM), or <em>finite state automaton</em> (FSA).</p>
<p>The flowchart above doesn't do a very good job of describing the FSM that our parser is, because it doesn't tell you <em>how</em> the machine transitions between states. The state <em>transitions</em> are what actually defines the format. For example, we know we are transitioning to a header when we see a <code>&gt;</code> symbol - that is, in a way, what <em>defines</em> the header state.</p>
<p>In the flowchart below, the arrows signifying transitions between states are marked with the input character(s) that causes the FSM (our parser) to transition, written in a <a href="http://regexr.com">regex-like</a> notation. For example, the transition from a header to a sequence is defined by a <code>\n</code> followed by one of the characters in <code>[ACGT]</code>. &ldquo;EOF&rdquo; signifies end of file. Note also that some states have transitions leading to itself. For example, when in the &ldquo;header&rdquo; state, the FSM may continue to read characters in <code>a-z</code> and stay in the &ldquo;header&rdquo; state.</p>
<pre><code class="language-julia">s = &quot;&quot;&quot;digraph {
  graph [ rankdir = LR ];
begin [ shape = circle ];
header [ shape = circle ];
sequence [ shape = circle ];
end [ shape = circle ];

begin -&gt; header [ label = &quot;&gt;[a-z]&quot; ];
header -&gt; header [ label = &quot;[a-z]&quot;]; 
header -&gt; sequence [ label = &quot;\\\\n[ACGT]&quot; ];
sequence -&gt; sequence [ label = &quot;[ACGT]&quot; ];
sequence -&gt; header [ label = &quot;\\\\n&gt;&quot; ];
sequence -&gt; end [ label = &quot;\\\\nEOF&quot; ];
begin -&gt; end;
}
&quot;&quot;&quot;
display_machine()
</code></pre>
<p><img src="automa_tutorial_4_0.svg" alt="svg"></p>
<p>How does the machine know when it's supposed to transition between states? Suppose a machine is at the beginning of a file and it begins by observing the input <code>&gt;</code>. Is this part of a transition to the header state, or the beginning of a malformed input?</p>
<p>It is certainly possible for the machine to have a <em>memory</em> of the pattern, such that it only transitions to the &ldquo;header&rdquo; state once it sees the two specific inputs <code>&gt;</code> and <code>[a-z]</code> in that order. However, we can make things much simpler by requiring the machine to make a state transition for <em>every single input byte</em>, then we don't need any memory other than what state the machine is in, which is a single integer. To still represent the same format at the flowchart above, we need to insert some extra states:</p>
<pre><code class="language-julia">s = &quot;&quot;&quot;digraph {
  graph [ rankdir = LR ];
  begin [ shape = circle ];
  2 [ shape = circle ];
1 [ shape = circle ];
3 [ shape = circle ];
  begin -&gt; 1 [ label = &quot;&gt;&quot; ];
  header [ shape = circle ];
1 -&gt; header [label = &quot;[a-z]&quot; ];
header -&gt; header [ label = &quot;[a-z]&quot; ];
2 -&gt; sequence [ label = &quot;[ACGT]&quot; ];
header -&gt; 2 [ label = &quot;\\\\n&quot; ];


sequence [ shape = circle ];
sequence -&gt; sequence [ label = &quot;[ACGT]&quot; ];
sequence -&gt; 3 [ label = &quot;\\\\n&quot; ];
3 -&gt; 1 [ label = &quot;&gt;&quot; ];
end [ shape = circle ];
3 -&gt; end [ label = &quot;EOF&quot; ];
&quot;begin&quot; -&gt; end [ label = &quot;EOF&quot; ];
}
&quot;&quot;&quot;
display_machine()
</code></pre>
<p><img src="automa_tutorial_6_0.svg" alt="svg"></p>
<p>The diagram above is exactly equivalent to the previous one, but this diagram makes a state transition for <em>every single</em> input byte. All bytes cause a transition, and no transition consumes multiple bytes. For simplicity, Automa's parsers work like that.</p>
<p>Note that it is always possible, given a FSM with multi-byte transitions to convert it to a FSM with single-byte transitions by just putting in more nodes. E.g. the transition from header to sequence requires two bytes, a newline and an A, C, G or T. Therefore, an intermediate note (on the illustration marked &ldquo;2&rdquo;) is needed. Unfortunately, the requirement of single-byte transitions limits what you can do with Automa - I'll come back to the limitations later.</p>
<p>Now let's create this same FSM using Automa.</p>
<h2 id="introducing-automa">Introducing Automa</h2>
<pre><code class="language-julia"># First imports
import Automa
import Automa.RegExp: @re_str
const re = Automa.RegExp;
</code></pre>
<p>We define the elements that make up the SimpleFasta (the header and the sequence) using <code>Automa.RegExp</code>, a subset of regular expressions.
Automa's regular expressions can be manipulated with the following operations:</p>
<table>
<thead>
<tr>
<th align="left">Function</th>
<th align="center">Alias</th>
<th align="left">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>cat(re...)</code></td>
<td align="center"><code>*</code></td>
<td align="left">concatenation</td>
</tr>
<tr>
<td align="left"><code>alt(re1, re2...)</code></td>
<td align="center"><code>\|</code></td>
<td align="left">alternation</td>
</tr>
<tr>
<td align="left"><code>rep(re)</code></td>
<td align="center"></td>
<td align="left">zero or more repetition</td>
</tr>
<tr>
<td align="left"><code>rep1(re)</code></td>
<td align="center"></td>
<td align="left">one or more repetition</td>
</tr>
<tr>
<td align="left"><code>opt(re)</code></td>
<td align="center"></td>
<td align="left">zero or one repetition</td>
</tr>
<tr>
<td align="left"><code>isec(re1, re2)</code></td>
<td align="center"><code>&amp;</code></td>
<td align="left">intersection</td>
</tr>
<tr>
<td align="left"><code>diff(re1, re2)</code></td>
<td align="center"><code>\</code>    | difference (subtraction) |</td>
<td></td>
</tr>
<tr>
<td align="left"><code>neg(re)</code></td>
<td align="center"><code>!</code></td>
<td align="left">negation</td>
</tr>
</tbody>
</table>
<p>Furthermore inside the re-strings themselves, you can use</p>
<ul>
<li>groups of characters, like <code>[ACGT]</code> representing the set <code>{A, C, G, T}</code></li>
<li>ranges <code>A-Z</code> representing <code>'A':'Z'</code> (or more accurately, the <em>bytes</em> <code>UInt8('A'):UInt8('Z')</code></li>
<li><code>*</code>, representing zero or more repetitions of the last element, such as <code>[a-z]*</code></li>
<li><code>+</code>, representing one or more repetitions of the last element, such as <code>[a-z]+</code></li>
<li><code>?</code>, representing zero or one repetition of the last element, such as <code>[a-z]?</code></li>
</ul>
<p>We can specify our FSM like below. We use a <code>let</code> statement only so we don't pollute global namespace with <code>header</code>, <code>sequence</code> etc.</p>
<pre><code class="language-julia">machine = let
    # Define SimpleFasta syntax
    header = re&quot;&gt;[a-z]+&quot;
    sequence = re&quot;[ACGT]+&quot;
    record = header * re&quot;\n&quot; * sequence * re&quot;\n&quot;
    fasta = re.rep(record)

    # Compile the regex to a FSM
    Automa.compile(fasta)
end;
</code></pre>
<p>If you have <code>dot</code> installed on your computer, you can visualize the compiled FSM. You might want to modify the paths in the function below to make it work on your computer.</p>
<pre><code class="language-julia">function display_machine(machine)
    write(&quot;/tmp/fasta_machine.dot&quot;, Automa.machine2dot(machine))
    run(`dot -Tsvg -o /tmp/fasta_machine.svg /tmp/fasta_machine.dot`)
    open(&quot;/tmp/fasta_machine.svg&quot;) do file
        display(&quot;image/svg+xml&quot;, String(read(file)))
    end
end;
</code></pre>
<pre><code class="language-julia">display_machine(machine)
</code></pre>
<p><img src="automa_tutorial_13_0.svg" alt="svg"></p>
<p>Yes, that looks correct!
The start node here is the small point on the left. You can see one of the states are represented by a double circle. This represents an <em>accept state</em>, where the machine may stop at a valid end of input - here after each SimpleFasta record. If the machine stops at any other state, it did not complete correctly. In other words, state &ldquo;1&rdquo; here doubles as the &ldquo;end&rdquo; state.</p>
<p>With this parser we can read a SimpleFasta file and determine if it conforms to the specification. In Automa, we do this by having our machine create Julia code in the form of a Julia expression (of type <code>Expr</code>), which we can then use to create a function using metaprogramming.</p>
<p>First, we need an <code>Automa.CodeGenContext</code>, an object which contains the options for code generation. For now, we don't worry about the settings and just make a default context object:</p>
<pre><code class="language-julia">context = Automa.CodeGenContext();
</code></pre>
<p>Then, we use two functions to generate code:</p>
<ul>
<li><code>Automa.generate_init_code(::CodeGenContext, ::Machine)</code> creates the Julia code needed to initialize the parsing of the given FSM.</li>
<li><code>Automa.generate_exec_code(::CodeGenContext, ::Machine, ::Dict{Symbol, Expr})</code> creates the Julia code needed for the main loop of the running parser. Here, the <code>Dict</code> is an <em>action dict</em>, which may contain arbitrary Julia code that the parser executes while parsing. We need that action dict later to make our parser do something useful - but for now, we simply want to create <em>a</em> parser, not necessarily a <em>useful</em> one, so we just use an empty dict. We will come back to the action dict later.</li>
</ul>
<p>Let's have a look at these functions:</p>
<pre><code class="language-julia">Automa.generate_init_code(context, machine)
</code></pre>
<pre><code>quote
    #= /home/jakob/.julia/packages/Automa/sfHZ6/src/codegen.jl:94 =#
    p::Int = 1
    #= /home/jakob/.julia/packages/Automa/sfHZ6/src/codegen.jl:95 =#
    p_end::Int = 0
    #= /home/jakob/.julia/packages/Automa/sfHZ6/src/codegen.jl:96 =#
    p_eof::Int = -1
    #= /home/jakob/.julia/packages/Automa/sfHZ6/src/codegen.jl:97 =#
    cs::Int = 1
end
</code></pre>
<p>Besides the comments, the init code seems to define four integers. What to they mean?</p>
<p>Automa works by reading the data byte by byte. When doing this, the data is represented in the function as an <code>AbstractVector{UInt8}</code> - let us call that <code>data</code>. <code>p</code> represents the index of <code>data</code> that the parser is currently reading from. <code>p_end</code> represents the last position of the data in <code>data</code>. Because <code>data</code> may just be a small, buffered part of a much larger file being parsed, <code>p_eof</code> represents the index of actual end of input. Last, <code>cs</code> is short for &ldquo;current state&rdquo;, and is an integer representing the state of the FSM. You can see that sensibly, it is initialized at state 1, and so, by looking at the FSM diagram, we can see it next expects a &lsquo;&gt;&rsquo;. The zero state is special, and represents the &ldquo;accept state&rdquo;, that is, the state of a correctly exited FSM.</p>
<p>The code generated by <code>Automa.generate_exec_code</code> is a little more complicated, and also depends on the <code>CodeGenContext</code>. But it will increment <code>p</code> by one, read a byte from <code>data</code>, and execute a state transition (i.e. change <code>cs</code>) depending on the byte. If <code>cs</code> goes negative, it knows the input is malformed. If <code>cs</code> is zero, the parser has finished. If <code>cs</code> is positive, it will continue until end of file.</p>
<p>Alright, let's create a parser using these functions!</p>
<pre><code class="language-julia">@eval function parse_fasta(data::Union{String,Vector{UInt8}})
    $(Automa.generate_init_code(context, machine))
    
    # p_end and p_eof were set to 0 and -1 in the init code,
    # we need to set them to the end of input, i.e. the length of `data`.
    p_end = p_eof = lastindex(data)
    
    # We just use an empty dict here because we don't want our parser
    # to do anything just yet - only parse the input
    $(Automa.generate_exec_code(context, machine, Dict{Symbol, Expr}()))

    # We need to make sure that we reached the accept state, else the 
    # input did not parse correctly
    iszero(cs) || error(&quot;failed to parse on byte &quot;, p)
    return nothing
end;
</code></pre>
<p>Does it work?</p>
<pre><code class="language-julia">for (inputno, data) in enumerate([
        &quot;&quot;,
        &quot;&gt;hi\nACGT\n&quot;,
        &quot;&gt;hi\nACGT&quot;,
        &quot;&gt;hi\nAA\n&gt;x\nGTGATCGTAGTA\n&quot;,
        &quot;&gt;hi\nthere!&quot;
    ])
    try
        parse_fasta(data)
        println(&quot;Input $inputno: parsed successfully!&quot;)
    catch e
        println(&quot;Input $inputno: &quot;, e)
    end
end
</code></pre>
<pre><code>Input 1: parsed successfully!
Input 2: parsed successfully!
Input 3: ErrorException(&quot;failed to parse on byte 9&quot;)
Input 4: parsed successfully!
Input 5: ErrorException(&quot;failed to parse on byte 5&quot;)
</code></pre>
<p>Excellent. It correctly parsed the empty data, input 2 and 4. Input 3 lacked a trailing newline, and input 5 didn't have a proper sequence.</p>
<p>But just parsing a file is not too useful. Normally, we parse a file in order to load it into some kind of datastructure to manipulate it. That's where the action dict comes into the picture. We can make Automa execute arbitrary Julia code during parsing. Here is how it works:</p>
<ul>
<li>When defining our machine, we add <em>action names</em>, in the form of Julia <code>Symbol</code>s to certain regexes. For example, we may define that exiting the regex <code>sequence</code> should execute the action <code>:exit_sequence</code>. Like other Julia identifiers, the name can be chosen arbitrarily.</li>
<li>Using the action dict, which was of type <code>Dict{Symbol, Expr}</code>, we can map action names to Julia code. This tells Automa what each action name <em>means</em>.</li>
<li>The actions expressions are put into the function we create at parse time, so there is no runtime cost to this.</li>
</ul>
<p>First, we redefine our machine. This time, we add actions:</p>
<pre><code class="language-julia">machine = let
    # Define SimpleFasta syntax
    newline = re&quot;\n&quot;
    header = re&quot;&gt;[a-z]+&quot;
    sequence = re&quot;[ACGT]+&quot;
    record = header * newline * sequence * newline
    fasta = re.rep(record)
        
    # Define SimpleFasta actions using arbitrary names
    header.actions[:enter] = [:enter]
    header.actions[:exit] = [:exit_header]
    sequence.actions[:enter] = [:enter]
    sequence.actions[:exit] = [:exit_sequence]
    record.actions[:exit] = [:exit_record]

    Automa.compile(fasta)
end;
</code></pre>
<p>If we visualize the machine now, the actions will be printed on the relevant edges in the FSM diagram. For example, the newline after a header is labeled <code>'\n'/exit_header</code>, meaning the transition will happen at the input byte <code>'\n'</code>, and it will execute the action <code>exit_header</code>.</p>
<p>Also note that the diagram structure itself changed, because the entering state and exit state are now distinct. The exit state 6 will execute <code>:exit_record</code> at the end of file (EOF), whereas if the FSM ends at state 1, <code>:exit_record</code> is not executed.</p>
<pre><code class="language-julia">display_machine(machine)
</code></pre>
<p><img src="automa_tutorial_25_0.svg" alt="svg"></p>
<p>Let's also define a simple data structure to parse the SimpleFasta records into. Like our SimpleFasta format, let's not overcomplicate things:</p>
<pre><code class="language-julia">using BioSequences
</code></pre>
<pre><code class="language-julia">struct FastaRecord{A}
    header::String
    sequence::LongSequence{A}
end

FastaRecord(h, sequence::LongSequence{A}) where A = FastaRecord{A}(h, sequence);
</code></pre>
<p>And now our action dict. I can use the variables <code>data</code>, <code>p</code>, <code>p_end</code> etc. in the code:</p>
<pre><code class="language-julia">actions = Dict(
    # Record which byte position marks the beginning of header or sequence
    :enter =&gt; quote
        mark = p
    end,
    :exit_header =&gt; quote
        header = String(data[mark+1:p-1])
    end,
    :exit_sequence =&gt; quote
        sequence = LongSequence{A}(data[mark:p-1])
    end,
    :exit_record =&gt; quote
        record = FastaRecord(header, sequence)
        push!(records, record)
    end,
);
</code></pre>
<pre><code class="language-julia">@eval function parse_fasta(::Type{A}, data::Union{String,Vector{UInt8}}) where {A &lt;: Alphabet}
    # We can't control the order that the code from out action dict is executed, so
    # we define variables beforehand so we can be sure they are not used before
    # they are defined.
    mark = 0
    records = FastaRecord{A}[]
    header = &quot;&quot;
    sequence = LongSequence{A}()

    # The rest is just like in our previous example, except now we use the
    # action dict we just created.
    $(Automa.generate_init_code(context, machine))
    p_end = p_eof = lastindex(data)
    $(Automa.generate_exec_code(context, machine, actions))

    iszero(cs) || error(&quot;failed to parse on byte &quot;, p)
    return records
end;
</code></pre>
<p>We can verify that it works:</p>
<pre><code class="language-julia">parse_fasta(DNAAlphabet{4}, &quot;&gt;hello\nTAG\n&gt;there\nTAA\n&quot;)
</code></pre>
<pre><code>2-element Array{FastaRecord{DNAAlphabet{4}},1}:
 FastaRecord{DNAAlphabet{4}}(&quot;hello&quot;, TAG)
 FastaRecord{DNAAlphabet{4}}(&quot;there&quot;, TAA)
</code></pre>
<h2 id="redesign-the-format">Redesign the format</h2>
<p>With the basics now covered, let's step up the game a tad and make it parse SimpleFasta records with multiple lines per sequence. We first define a <code>seqline</code> pattern that looks like the old <code>sequence</code> pattern, and build a new <code>sequence</code> pattern from multiple <code>seqline</code> patterns.</p>
<pre><code class="language-julia">machine = let
    # Define SimpleFasta syntax
    header = re&quot;&gt;[a-z]+&quot;
    seqline = re&quot;[ACGT]+&quot;
    sequence = seqline * re.rep(re&quot;\n&quot; * seqline)
    record = header * re&quot;\n&quot; * sequence * re&quot;\n&quot;
    fasta = re.rep(record)
        
    # Define SimpleFasta actions
    header.actions[:enter] = [:enter]
    header.actions[:exit] = [:exit_header]
    seqline.actions[:enter] = [:enter]
    seqline.actions[:exit] = [:exit_seqline]
    record.actions[:exit] = [:exit_record]

    Automa.compile(fasta)
end;
</code></pre>
<p>The machine now has a subtle change with a small loop between node 5 and 6 representing the parser seeing multiple sequence lines.</p>
<pre><code class="language-julia">display_machine(machine)
</code></pre>
<p><img src="automa_tutorial_37_0.svg" alt="svg"></p>
<p>We also need to update the <code>actions</code>. Here, I use <code>:(this syntax)</code>, which is equivalent to</p>
<pre><code>quote 
    this syntax
end
</code></pre>
<p>but more readable for one-liners. I keep a buffer containing the sequence, and append every sequence line to the buffer at the end of each <code>seqline</code>. When the record is done, I create the sequence and empty the buffer for the next record.</p>
<pre><code class="language-julia">actions = Dict(
    :enter =&gt; :(mark= p),
    :exit_header =&gt; :(header = String(data[mark+1:p-1])),
    :exit_seqline =&gt; quote
        doff = length(seqbuffer) + 1
        resize!(seqbuffer, length(seqbuffer) + p - mark)
        copyto!(seqbuffer, doff, data, mark, p-mark)
    end,
    :exit_record =&gt; quote
        sequence = LongSequence{A}(seqbuffer)
        empty!(seqbuffer)
        record = FastaRecord(header, sequence)
        push!(records, record)
    end,
);
</code></pre>
<p>We need to redefine <code>parse_fasta</code>, now also containing the <code>seqbuffer</code>:</p>
<pre><code class="language-julia">@eval function parse_fasta(::Type{A}, data::Union{String,Vector{UInt8}}) where {A &lt;: Alphabet}
    mark = 0
    records = FastaRecord{A}[]
    header = &quot;&quot;
    sequence = LongSequence{A}()
    seqbuffer = UInt8[]

    $(Automa.generate_init_code(context, machine))
    p_end = p_eof = lastindex(data)
    $(Automa.generate_exec_code(context, machine, actions))

    iszero(cs) || error(&quot;failed to parse on byte &quot;, p)
    return records
end;
</code></pre>
<p>And we can confirm it now works for multiple sequences!</p>
<pre><code class="language-julia">parse_fasta(DNAAlphabet{2}, &quot;&gt;hello\nTAG\nGGG\n&gt;there\nTAA\nTAG\n&quot;)
</code></pre>
<pre><code>2-element Array{FastaRecord{DNAAlphabet{2}},1}:
 FastaRecord{DNAAlphabet{2}}(&quot;hello&quot;, TAGGGG)
 FastaRecord{DNAAlphabet{2}}(&quot;there&quot;, TAATAG)
</code></pre>
<h2 id="even-more-complexity">Even more complexity</h2>
<p>One of the greatest things about Automa is how we can parse quite complicated formats without needing to manually construct much code. For example, here, I change <em>only</em> the regular expressions:</p>
<ul>
<li>I allow optional Windows-style newlines (<code>\r\n</code>)</li>
<li>I allow trailing whitespace on each line (<code>re.space() \ (re&quot;\r&quot; | re&quot;\n&quot;)</code> means all space characters except \r or \n), but not whitespace inside headers</li>
<li>Sequences can only contain all UIPAC ambiguous RNA or DNA nucleotides</li>
<li>The trailing record no longer needs a trailing newline to be valid. It has the same actions as a regular record.</li>
</ul>
<pre><code class="language-julia">machine = let
    # Define FASTA syntax
    newline = re.opt(&quot;\r&quot;) * re&quot;\n&quot;
    hspace = re.space() \ (re&quot;\r&quot; | re&quot;\n&quot;)
    header = re&quot;&gt;&quot; * re.rep1((re.any() \ re.space())) * re.opt(hspace)
    seqline = re&quot;[acgturmkyswbdhvnACGTURMKYSWBDHVN\-*]+&quot; * re.opt(hspace)
    sequence = seqline * re.rep(newline * seqline)
    record = header * newline * sequence * newline
    trailing_record = header * newline * sequence * re.rep(newline * re.opt(hspace))
    fasta = re.rep(newline) * re.rep(record) * re.opt(trailing_record)
        
    # Define FASTA actions
    header.actions[:enter] = [:enter]
    header.actions[:exit] = [:exit_header]
    seqline.actions[:enter] = [:enter]
    seqline.actions[:exit] = [:exit_seqline]
    record.actions[:exit] = [:exit_record]
    trailing_record.actions[:exit] = [:exit_record]

    Automa.compile(fasta)
end;
</code></pre>
<p>The machine is now a bit more complicated. But who cares, it's automatically generated. Look how easy it was to generate!</p>
<pre><code class="language-julia">display_machine(machine)
</code></pre>
<p><img src="automa_tutorial_47_0.svg" alt="svg"></p>
<h2 id="pitfall-1-ambiguous-parsers">Pitfall 1: Ambiguous parsers</h2>
<p>Besides its complexity, building parsers with Automa has some pitfalls - notably those caused by Automa transitioning state for every input byte.</p>
<p>Unfortunately, it's very easy to accidentally create a machine that can't possibly figure out what action to take when looking only at one byte at a time. The following simple machine will not compile (on the master branch of Automa).</p>
<pre><code class="language-julia">bad_machine = let
    left = re&quot;Turn left!&quot;
    right = re&quot;Turn right!&quot;
    left_or_right = left | right
    
    left.actions[:enter] = [:turn_left]
    right.actions[:enter] = [:turn_right]
    
    Automa.compile(left_or_right)
end
</code></pre>
<pre><code>Ambiguous DFA: Input 0x54 can lead to actions [:turn_right] or [:turn_left]



Stacktrace:

 [1] error(::String) at ./error.jl:33

 [2] validate_paths(::Array{Tuple{Union{Nothing, Automa.Edge},Array{Symbol,1}},1}) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:176

 [3] validate_nfanodes(::Automa.StableSet{Automa.NFANode}) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:195

 [4] nfa2dfa(::Automa.NFA, ::Bool) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:104

 [5] compile(::Automa.RegExp.RE; optimize::Bool, unambiguous::Bool) at /home/jakob/.julia/packages/Automa/sfHZ6/src/machine.jl:55

 [6] compile(::Automa.RegExp.RE) at /home/jakob/.julia/packages/Automa/sfHZ6/src/machine.jl:55

 [7] top-level scope at In[27]:9

 [8] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091
</code></pre>
<p>Automa refuse to create this FSM. Why? Well, the first byte it expects is a <code>T</code> (or <code>0x54</code>) - but there is no way of knowing whether it's entering <code>left</code> or <code>right</code> when it sees that byte, and these two patters have conflicting actions! If they had the same actions, there would be no problem.</p>
<p>This situation is highly artificial, but the situation happens often in real life. Here's a very subtle change to one of the previous parsers:</p>
<pre><code class="language-julia">bad_machine = let
    # Define SimpleFasta syntax
    header = re&quot;&gt;[a-z]+&quot;
    seqline = re&quot;[ACGT]+&quot;
    sequence = seqline * re.rep(re&quot;\n&quot; * seqline)
    record = header * re&quot;\n&quot; * sequence
    fasta = re.rep(record * re&quot;\n&quot;)
        
    # Define SimpleFasta actions
    header.actions[:enter] = [:enter]
    header.actions[:exit] = [:exit_header]
    seqline.actions[:enter] = [:enter]
    seqline.actions[:exit] = [:exit_seqline]
    record.actions[:exit] = [:exit_record]

    Automa.compile(fasta)
end;
</code></pre>
<pre><code>Ambiguous DFA: Input 0x0a can lead to actions [:exit_seqline, :exit_record] or [:exit_seqline]



Stacktrace:

 [1] error(::String) at ./error.jl:33

 [2] validate_paths(::Array{Tuple{Union{Nothing, Automa.Edge},Array{Symbol,1}},1}) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:176

 [3] validate_nfanodes(::Automa.StableSet{Automa.NFANode}) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:195

 [4] nfa2dfa(::Automa.NFA, ::Bool) at /home/jakob/.julia/packages/Automa/sfHZ6/src/dfa.jl:104

 [5] compile(::Automa.RegExp.RE; optimize::Bool, unambiguous::Bool) at /home/jakob/.julia/packages/Automa/sfHZ6/src/machine.jl:55

 [6] compile(::Automa.RegExp.RE) at /home/jakob/.julia/packages/Automa/sfHZ6/src/machine.jl:55

 [7] top-level scope at In[28]:16

 [8] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091
</code></pre>
<p>After a <code>seqline</code>, when encountering a newline, there is no way of knowing whether this marks the end of a record or whether it continues at the next line. Automa can't look ahead, it has to make a decision at every byte.</p>
<p>Here, the solution is to move the newline from the definition of <code>fasta</code> to that of <code>sequence</code>. That way, the newline functions as a kind of one-byte lookahead - if the byte after the newline is a <code>&gt;</code>, it knows the record is complete.</p>
<p>In general, encountering situations like this is usually a sign that the parser is badly built - or that the format is. You can usually resolve it by using single-byte lookahead like above. If not, it is possible to optionally disable the check for ambiguous actions by passing a keyword to the <code>Automa.parse</code> function. But beware that this may lead to absurd results.</p>
<h2 id="pitfall-2-no-recursively-defined-patterns">Pitfall 2: No recursively defined patterns</h2>
<p>Some formats are naturally recursive. The <a href="https://en.wikipedia.org/wiki/Newick_format">Newick format</a>, for example, defines &ldquo;subtree&rdquo; in terms of &ldquo;internal&rdquo;, which is defined in terms of &ldquo;branchset&rdquo;, defined in terms of &ldquo;branch&rdquo; which is defined in terms of&hellip; &ldquo;subtree&rdquo;.</p>
<p>Something like that will not work Automa:</p>
<pre><code class="language-julia">simple_newick = let
    name = re&quot;[a-z_]&quot;
    clade = name | (re&quot;\(&quot; * cladeset * &quot;)&quot;)
    cladeset = clade * re.rep(re&quot;,&quot; * clade)
    Automa.compile(cladeset * re&quot;;&quot;)
end
    
</code></pre>
<pre><code>UndefVarError: cladeset not defined



Stacktrace:

 [1] top-level scope at In[29]:3

 [2] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091
</code></pre>
<p>Because, in general, you can't refer to objects in Julia that have not yet been defined. Worse, even if the syntactical challenge was solved, recursive patterns usually only make sense if you have lookahead. Newick files, for example, simply <em>can't</em> be parsed by FSMs, because every time you see an open parenthesis, you need to make sure there is a closing one further in the file, and this requires lookahead that can't be resolved by parsing one byte at a time.</p>
<p>Luckily, because Automa can execute arbitrary Julia code, we can have the FSM manipulate a stack and turn the FSM into a <em>pushdown automaton</em>, which can handle formats like Newick. It does require a little more manual fiddling, and is less elegant:</p>
<pre><code class="language-julia">simple_newick = let
    name = re&quot;[a-z_]+&quot;
    cladestart = re&quot;\(&quot;
    cladestop = re&quot;\)&quot;
    cladesep = re&quot;,&quot;
    nodechange = cladestart | cladestop | cladesep
    newick_element = nodechange * re.opt(name)
    tree = re.opt(name) * re.rep(newick_element) * re&quot;;&quot;
    
    cladestart.actions[:enter] = [:push]
    cladestop.actions[:enter] = [:pop]
    cladesep.actions[:enter] = [:pop, :push]
    
    Automa.compile(tree)
end 

actions = Dict(
    :push =&gt; quote
        level -= 1
    end,
    :pop =&gt; quote
        iszero(level) &amp;&amp; error(&quot;Too many closing parentheses&quot;)
        level += 1
    end
);
</code></pre>
<pre><code class="language-julia">@eval function parse_newick(data::Union{String,Vector{UInt8}})
    level = 0
    $(Automa.generate_init_code(context, simple_newick))
    p_end = p_eof = lastindex(data)
    $(Automa.generate_exec_code(context, simple_newick, actions))

    iszero(cs) || error(&quot;failed to parse on byte &quot;, p)
    iszero(level) || error(&quot;Too many opening parentheses&quot;)
end;
</code></pre>
<p>And it sort of works:</p>
<pre><code class="language-julia">println(&quot;Does this parse? &quot;, parse_newick(&quot;(hello,hi);&quot;))
println(&quot;Does this parse? &quot;, parse_newick(&quot;(hello,hi)(;&quot;))
</code></pre>
<pre><code>Does this parse? true



Too many opening parentheses



Stacktrace:

 [1] error(::String) at ./error.jl:33

 [2] parse_newick(::String) at ./In[31]:8

 [3] top-level scope at In[32]:2

 [4] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091
</code></pre>
<p>Here, I needed to manually keep track of the level of nesting. In fact, I would have to keep track of more stuff to disallow inputs like <code>();</code>, removing some of the point of Automa - namely, that it is automatic.</p>
<p>In the next part of this tutorial, I will show how to use Automa to create honest-to-God parsers which integrate with the BioJulia ecosystem and can work on streamed data.</p>
<h2 id="optimizing">Optimizing</h2>
<p>Automa's parsers are pretty fast. To create hand-written parsers of comparable speeds, you need to microoptimize every operation, which is annoying. That being said, for actually loading parsed files to be fast, you need to optimize the user-defined actions in the actions dictionary, too.</p>
<p>How fast is our un-optimized Fasta parser already? Let's test it on some data and find out. I'll make 50k sequences with 2k base pairs each, for a total of 100 MB data.</p>
<p>Remember to re-run the block of code where the <code>parse_fasta</code> function is defined since we changed the machine.</p>
<pre><code class="language-julia"># Create 50k records with 2kbp each
function generate_data()
    open(&quot;/tmp/fasta.fna&quot;, &quot;w&quot;) do file
        for seq in 1:50000
            println(file, '&gt;', join(rand('a':'z', 16)))
            for line in 1:20
                println(file, join(rand(&quot;ACGT&quot;, 100)))
            end
        end
    end
end
generate_data()
</code></pre>
<pre><code class="language-julia">function parsedata()
    open(&quot;/tmp/fasta.fna&quot;) do file
        parse_fasta(DNAAlphabet{2}, read(file))
    end
end;
</code></pre>
<pre><code class="language-julia">parsedata();
@time parsedata();
</code></pre>
<pre><code>  0.330489 seconds (200.04 k allocations: 138.861 MiB, 4.15% gc time)
</code></pre>
<p>Alright! It does about 300 MB/s on my laptop. Not bad! I'd say for most applications, this speed is more than sufficient.</p>
<p>But this is Julia. We want to be able to optimize the crap out of it. So let's optimize the actions:</p>
<pre><code class="language-julia">actions = Dict(
    :enter =&gt; :(mark = p),
    
    # Create string with only one copy of the data.
    :exit_header =&gt; :(header = unsafe_string(pointer(data, mark + 1), p - mark - 1)),
    
    # Only resize buffer if it's too small, else just keep track of how many bytes are used.
    :exit_seqline =&gt; quote
        N = p - mark
        length(seqbuffer) &lt; filled + N &amp;&amp; resize!(seqbuffer, filled + N)
        copyto!(seqbuffer, filled+1, data, mark, N)
        filled += N
    end,
    
    # Use `copyto!` to only encode data directly from data vector into sequence
    # note this requires the latest versions of BioSequences.
    :exit_record =&gt; quote
        sequence = copyto!(LongSequence{A}(filled), 1, seqbuffer, 1, filled)
        record = FastaRecord(header, sequence)
        push!(records, record)
        filled = 0
    end,
);
</code></pre>
<p>If we want to optimize, we also need to use the fastest <code>CodeGenContext</code>. We use the <code>:goto</code>-generator. This creates machine code using <code>@goto</code>-statements, which creates very long and completely unreadable code - but it's fast. Also, who cares if it's hard to read. It's machine-generated code. We disable the FSM's boundschecks when accessing the <code>data</code> vector - since we don't manipulate the position <code>p</code> manually, we are confident it will not go out of bounds. And we allow the code generator to unroll loops:</p>
<pre><code class="language-julia">context = Automa.CodeGenContext(generator=:goto, checkbounds=false, loopunroll=4);
</code></pre>
<p>We also need to modify the <code>parse</code> function because we use a new variable called <code>filled</code> and need to initialize it.</p>
<pre><code class="language-julia">@eval function parse_fasta(::Type{A}, data::Union{String,Vector{UInt8}}) where {A &lt;: Alphabet}
    mark = 0
    records = FastaRecord{A}[]
    header = &quot;&quot;
    sequence = LongSequence{A}()
    seqbuffer = UInt8[]
    filled = 0

    $(Automa.generate_init_code(context, machine))
    p_end = p_eof = lastindex(data)
    $(Automa.generate_exec_code(context, machine, actions))

    iszero(cs) || error(&quot;failed to parse on byte &quot;, p)
    return records
end;
</code></pre>
<pre><code class="language-julia">parsedata();
@time parsedata();
</code></pre>
<pre><code>  0.162113 seconds (150.04 k allocations: 133.520 MiB, 7.90% gc time)
</code></pre>
<p>Okay, we're at slightly above 600 MB/s. Profiling confirms nearly all time is spent encoding the DNA sequences to <code>LongSequence</code>. We can make it faster still by not storing the data as a <code>LongSequence</code>, and instead make it a &ldquo;lazy&rdquo; object that only constructs the sequence upon demand. Even further, we could avoid heap-allocating each sequence individually. But these optimizations do not pertain directly to Automa, and I will leave them here.</p>

    </div>

    







<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://BioJulia.github.io/post/automa1/&amp;text=Tutorial%20to%20Automa:%20Part%201" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://BioJulia.github.io/post/automa1/&amp;t=Tutorial%20to%20Automa:%20Part%201" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Tutorial%20to%20Automa:%20Part%201&amp;body=https://BioJulia.github.io/post/automa1/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://BioJulia.github.io/post/automa1/&amp;title=Tutorial%20to%20Automa:%20Part%201" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://web.whatsapp.com/send?text=Tutorial%20to%20Automa:%20Part%201%20https://BioJulia.github.io/post/automa1/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://BioJulia.github.io/post/automa1/&amp;title=Tutorial%20to%20Automa:%20Part%201" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>












  
  
    
  
  






  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <img class="portrait mr-3" src="/authors/jakobnissen/avatar_hu4ad62102e3705c78ff7637ab3f43ebb3_23980_250x250_fill_q90_lanczos_center.jpg" alt="Avatar">
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/jakobnissen/">Jakob Nybo Nissen</a></h5>
      <h6 class="card-subtitle">BioJulia Developer</h6>
      <p class="card-text">Bioinformatician at Statens Serum Institut working with pandemic influenza genomics</p>
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:jakobnybonissen@gmail.com" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/jakobnissen" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>




<section id="comments">
  
    
<div id="disqus_thread"></div>
<script>
  let disqus_config = function () {
    
    
    
  };
  (function() {
    if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
      document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
      return;
    }
    var d = document, s = d.createElement('script'); s.async = true;
    s.src = 'https://' + "biojulia" + '.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


  
</section>






  
  



  </div>
</article>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/highlight.min.js" integrity="sha256-1zu+3BnLYV9LdiY85uXMzii3bdrkelyp37e0ZyTAQh0=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/julia.min.js"></script>
        
      

    

    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.5.1/leaflet.js" integrity="sha256-EErZamuLefUnbMBQbsEqu1USa+btR2oIlCpBJbyD4/g=" crossorigin="anonymous"></script>
    

    
    
    <script>const code_highlighting = true;</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    <script id="dsq-count-scr" src="https://biojulia.disqus.com/count.js" async></script>
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.a0d331bcd05dbe8b31e244f796710f08.js"></script>

    






  
  
  <div class="container">
    <footer class="site-footer">
  

  <p class="powered-by">
    Sabrina Ward &middot; 

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" class="back-to-top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
